### PR TITLE
Add taxonomy block using SelectTree

### PR DESCRIPTION
**Pull request by nathanss**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This creates a "Hierarchical Taxonomy" block, which will be firstly used for WooCommerce Brands plugin, and also replace woocommerce/product-category-field block in the future.

Part of https://github.com/woocommerce/woocommerce/issues/39206 .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install the "Custom Post Type UI" plugin by WebDevStudios
2. Set up some custom taxonomies.
    - Make sure to attach it to the Products post type
3. Add code similar to the following to a plugin/mu-plugin file, to add the properties to the REST API response and request:

```php
function YOUR_PREFIX_rest_api_prepare_custom1_to_product( $response, $post ) {
	$post_id = $post->get_id();

	if ( empty( $response->data[ 'custom1' ] ) ) {
		$terms = [];

		foreach ( wp_get_post_terms( $post_id, 'custom-taxonomy' ) as $term ) {
			$terms[] = [
				'id'   => $term->term_id,
				'name' => $term->name,
				'slug' => $term->slug,
			];
		}

		$response->data[ 'custom1' ] = $terms;
	}

	return $response;
}

add_filter( 'woocommerce_rest_prepare_product_object', 'YOUR_PREFIX_rest_api_prepare_custom1_to_product', 10, 2 );

function YOUR_PREFIX_rest_api_add_custom1_to_product( $product, $request, $creating = true ) {
	$product_id = $product->get_id();
	$params     = $request->get_params();
	$custom1s     = isset( $params['custom1'] ) ? $params['custom1'] : array();

	if ( ! empty( $custom1s ) ) {
		if ( $custom1s[0]['id'] ) {
			$custom1s = array_map(
				function ( $custom1 ) {
					return absint( $custom1['id'] );
				},
				$custom1s
			);
		} else {
			$custom1s = array_map( 'absint', $custom1s );
		}
		wp_set_object_terms( $product_id, $custom1s, 'custom-taxonomy' );
	}
}

add_filter( 'woocommerce_rest_insert_product_object', 'YOUR_PREFIX_rest_api_add_custom1_to_product', 10, 3 );
```

4. Add a block for the taxonomy property, using code similar to the following:
```php
$section_fields[] = [
	'woocommerce/taxonomy-field',
	[
		'slug'               => 'custom-taxonomy',
		'property' => 'custom1',
		'label'                                  => __( 'Custom Taxonomy', 'woocommerce' ),
		'createTitle'                        => __( 'Create new custom taxonomy', 'woocommerce' ),
	]
];
```

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: packages/js/product-editor/src/blocks/hierarchical-taxonomy/edit.tsx [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	);
+
+	useEffect( () => {
+		searchDelayed( '' );
+	}, [] );
+
+	const [ selectedEntries, setSelectedEntries ] = useEntityProp< Taxonomy[] >(
+		'postType',
+		'product',
+		taxonomyNamePlural
```
Comment (mattsherman) [inline packages/js/product-editor/src/blocks/hierarchical-taxonomy/edit.tsx]: I think this is actually not necessarily the plural version of the taxonomy name, but rather the name of the property used in the REST API.

File: packages/js/product-editor/src/blocks/hierarchical-taxonomy/editor.scss [hunk 2]
```diff
@@ -0,0 +1,16 @@
+.components-modal__screen-overlay {
+	.woocommerce-create-new-taxonomy-modal {
+		min-width: 650px;
+		overflow: visible;
+
+		&__buttons {
+			margin-top: 36px;
+			display: flex;
+			flex-direction: row;
+			gap: 8px;
```
Comment (mattsherman) [inline packages/js/product-editor/src/blocks/hierarchical-taxonomy/editor.scss]: Can use `$gap-smaller`

File: packages/js/product-editor/src/blocks/hierarchical-taxonomy/editor.scss [hunk 3]
```diff
@@ -0,0 +1,16 @@
+.components-modal__screen-overlay {
+	.woocommerce-create-new-taxonomy-modal {
+		min-width: 650px;
+		overflow: visible;
+
+		&__buttons {
+			margin-top: 36px;
```
Comment (mattsherman) [inline packages/js/product-editor/src/blocks/hierarchical-taxonomy/editor.scss]: Can use `$gap-larger`

**Review by mattsherman — CHANGES_REQUESTED** [review]
Nice start on this @nathanss !

I need to take a closer look, but some initial thoughts...

* Some testing instructions that don't depend on the brands PR/plugin would be good
* Relatedly, it would be good to verify that this works with other custom taxonomies (I tried creating some with the "Custom Post Type UI" plugin, but I don't know why they don't seem to be appearing in the REST API responses (which I think might be the problem).
* It would be ideal if this block automatically handled both hierarchical and non-hierarchical taxonomies. I would think that should be possible without too much hassle.

**Comment by nathanss** [comment]
@mattsherman thx for the review!

> Some testing instructions that don't depend on the brands PR/plugin would be good

I can't think of a good way of doing this. I don't know how using `useEntityProp` would work in another context, like storybook, for example. One other way would be to migrate the category field. But ultimately I think the best way of testing this is with the brands plugin for now, since it is the first use case. Do you have any suggestions?

> Relatedly, it would be good to verify that this works with other custom taxonomies (I tried creating some with the "Custom Post Type UI" plugin, but I don't know why they don't seem to be appearing in the REST API responses (which I think might be the problem).

You're right, we need to have the property in the REST API. AFAIK we don't have any other property on the product to test this on for now.

> It would be ideal if this block automatically handled both hierarchical and non-hierarchical taxonomies. I would think that should be possible without too much hassle.

I will work on that.

**Comment by mattsherman** [comment]
> @mattsherman thx for the review!
> 
> > Some testing instructions that don't depend on the brands PR/plugin would be good
> 
> I can't think of a good way of doing this. I don't know how using `useEntityProp` would work in another context, like storybook, for example. One other way would be to migrate the category field. But ultimately I think the best way of testing this is with the brands plugin for now, since it is the first use case. Do you have any suggestions?

Ideally, I think testing one PR should not depend on changes in another PR. I have figured out how to test this without the Brands plugin (which is needed to test the non-hierarchical case too).

> > Relatedly, it would be good to verify that this works with other custom taxonomies (I tried creating some with the "Custom Post Type UI" plugin, but I don't know why they don't seem to be appearing in the REST API responses (which I think might be the problem).
> 
> You're right, we need to have the property in the REST API. AFAIK we don't have any other property on the product to test this on for now.

I figured out how to accomplish this.

> > It would be ideal if this block automatically handled both hierarchical and non-hierarchical taxonomies. I would think that should be possible without too much hassle.
> 
> I will work on that.

Thanks! 

Okay, here is an approach I have figured out to test this easily...

1. Install the "Custom Post Type UI" plugin by WebDevStudios (it's free; you really don't need it though... you could just register a taxonomy via code... I just used this to save a little bit of time/typing/learning when trying to get things working)
2. Set up some custom taxonomies.
    - "Taxonomy Slug" is our "taxonomyName"
    - Make sure to attach it to the Products post type
3. Add code similar to the following to a plugin/mu-plugin file, to add the properties to the REST API response:

```php
function YOUR_PREFIX_rest_api_add_custom1_to_product( $response, $post ) {
	$post_id = $post->get_id();

	if ( empty( $response->data[ 'custom1' ] ) ) {
		$terms = [];

		foreach ( wp_get_post_terms( $post_id, 'custom-taxonomy' ) as $term ) {
			$terms[] = [
				'id'   => $term->term_id,
				'name' => $term->name,
				'slug' => $term->slug,
			];
		}

		$response->data[ 'custom1' ] = $terms;
	}

	return $response;
}

add_filter( 'woocommerce_rest_prepare_product_object', 'YOUR_PREFIX_rest_api_add_custom1_to_product', 10, 2 );
```

4. Add a block for the taxonomy property, using code similar to the following (using pre-API technique just because we don't have template registration in quite yet):

```php
$section_fields[] = [
	'woocommerce/taxonomy-field',
	[
		'taxonomyName'               => 'custom-taxonomy',
		'taxonomyNameOnEntity' => 'custom1',
		'label'                                  => __( 'Custom Taxonomy', 'woocommerce' ),
		'createTitle'                        => __( 'Create new custom taxonomy', 'woocommerce' ),
	]
];
```

File: packages/js/product-editor/src/blocks/taxonomy/block.json [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	"category": "widgets",
+	"description": "A block that displays a taxonomy field, allowing to search, select, and create new entries",
+	"keywords": [ "taxonomy"],
+	"textdomain": "default",
+	"attributes": {
+		"taxonomyName": {
+			"type": "string",
+			"__experimentalRole": "content"
+		},
+		"taxonomyNameOnEntity": {
```
Comment (mattsherman) [inline packages/js/product-editor/src/blocks/taxonomy/block.json]: How about `property`? That's what other generic fields like our checkbox and radio button blocks use.

File: packages/js/product-editor/src/blocks/taxonomy/block.json [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+		},
+		"createTitle": {
+			"type": "string",
+			"__experimentalRole": "content"
+		},
+		"hierarchical": {
+			"type": "boolean",
+			"default": false,
+			"__experimentalRole": "content"
+		}
```
Comment (mattsherman) [inline packages/js/product-editor/src/blocks/taxonomy/block.json]: I don't think the consumer should set this. We can determine this by querying for details on the taxonomy.

File: packages/js/product-editor/src/blocks/taxonomy/block.json [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+	"$schema": "https://schemas.wp.org/trunk/block.json",
+	"apiVersion": 2,
+	"name": "woocommerce/taxonomy-field",
+	"title": "Taxonomy",
+	"category": "widgets",
+	"description": "A block that displays a taxonomy field, allowing to search, select, and create new entries",
+	"keywords": [ "taxonomy"],
+	"textdomain": "default",
+	"attributes": {
+		"taxonomyName": {
```
Comment (mattsherman) [inline packages/js/product-editor/src/blocks/taxonomy/block.json]: This is the taxonomy slug. Maybe we should call it `taxonomySlug`.

**Review by mattsherman — CHANGES_REQUESTED** [review]
Excellent iteration on this so far! Overall, things appear to be working pretty well with the latest changes. I will do a more rigorous final test and walk through the code once the remaining changes are made.

**Comment by nathanss** [comment]
@mattsherman Awesome review, thanks a lot.

I updated the code with the suggested changes. I also updated the PR title and description, since it evolved a lot since the beginning. Can you check again?

File: packages/js/product-editor/src/blocks/taxonomy/block.json [hunk 7]
```diff
@@ -0,0 +1,38 @@
+{
+	"$schema": "https://schemas.wp.org/trunk/block.json",
+	"apiVersion": 2,
+	"name": "woocommerce/taxonomy-field",
+	"title": "Taxonomy",
+	"category": "widgets",
+	"description": "A block that displays a taxonomy field, allowing to search, select, and create new entries",
```
Comment (mattsherman) [inline packages/js/product-editor/src/blocks/taxonomy/block.json]: Suggestion...

```
"A block that displays a taxonomy field, allowing searching, selection, and creation of new items"
```

File: packages/js/product-editor/src/blocks/taxonomy/create-taxonomy-modal.tsx [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+					} ) }
+				>
+					<TextControl
+						id={ id }
+						value={ name }
+						onChange={ setName }
+					/>
+				</BaseControl>
+				{ hierarchical && (
+					<SelectTree
```
Comment (mattsherman) [inline packages/js/product-editor/src/blocks/taxonomy/create-taxonomy-modal.tsx:126]: This select tree does not show parents nested, as the categories modal does. It shows them in a flat list, which is confusing....

Screenshot 2023-08-30 at 19 53 31

Screenshot 2023-08-30 at 19 54 20

File: packages/js/product-editor/src/blocks/taxonomy/edit.tsx [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+export function Edit( {
+	attributes,
+}: {
+	attributes: TaxonomyBlockAttributes;
+} ) {
+	const blockProps = useBlockProps();
+	const hierarchical = useSelect( ( select ) => {
+		const taxonomyDetails = select( 'core' ).getTaxonomy(
+			attributes.slug
+			// eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (mattsherman) [inline packages/js/product-editor/src/blocks/taxonomy/edit.tsx]: Can we avoid this?

**Comment by mattsherman** [comment]
Note: forgot that we have to handle setting the taxonomy in the REST request as well... updated the testing instructions to account for that.

**Review by mattsherman — CHANGES_REQUESTED** [review]
Testing seems to work pretty well with the `brands` taxonomy, as well as two custom taxonomies I created, one hierarchical, one flat (after I added the handling to save the taxonomy in the REST handler*).

The main issue I ran into during testing was that the parent dropdown in the modal does not show parents hierarchically.

\* I suppose that this generic taxonomy block only works out of the box if the custom taxonomy developer has implemented support for the REST API as is being done for brands and categories (same shape of data for the property). We'll have to document this. I think it is a reasonable limitation, and maybe one we can extend over time if there are other patterns.

**Review by mattsherman — APPROVED** [review]
Nice work on this, @nathanss -- latest changes look good! :shipit:

**Pull request merged**